File: Chapter- 11/lee_algo.cpp
#include <bits/stdc++.h>
using namespace std;
#define ROW 5
#define COL 5

struct Cell
{
	int x;
	int y;
};

struct queue_Node
{
	Cell pt;
	int dist;
};

bool check_cellValid(int row, int col)
{
	return ((row >= 0) && (row < ROW) &&	(col >= 0) && (col < COL));
}

int rowNum[] = {-1, 0, 0, 1};
int colNum[] = {0, -1, 1, 0};

int Lee_algo(int matrix[][COL], Cell src, Cell dest)
{
	if (!matrix[src.x][src.y] || !matrix[dest.x][dest.y])
		return -1;

	bool visited[ROW][COL];
	memset(visited, false, sizeof visited);

	visited[src.x][src.y] = true;

	queue<queue_Node> q;


	queue_Node s = {src, 0};
	q.push(s);

	while (!q.empty())
	{
		queue_Node curr = q.front();
		Cell pt = curr.pt;

		if (pt.x == dest.x && pt.y == dest.y)
			return curr.dist;
		q.pop();

		for (int i = 0; i < 4; i++)
		{
			int row = pt.x + rowNum[i];
			int col = pt.y + colNum[i];


			if (check_cellValid(row, col) && matrix[row][col] &&
			!visited[row][col])
			{

				visited[row][col] = true;
				queue_Node adj_Cell = { {row, col},
									curr.dist + 1 };
				q.push(adj_Cell);
			}
		}
	}

	return -1;
}

int main()
{
	int matrix[ROW][COL] =
	{
		{ 1, 0, 0, 0, 1 },
		{ 1, 0, 1, 0, 1 },
		{ 1, 0, 1, 0, 1 },
		{ 0, 0, 0, 0, 0 },
		{ 0, 1, 0, 0, 1 }};

	Cell source = {0, 0};
	Cell dest = {2, 1};

	int dist = Lee_algo(matrix, source, dest);

	if (dist != -1)
		cout<< "Shortest Path is : "<<dist ;
	else
		cout<< "Shortest Path doesn't exist ";

	return 0;
}

File: Chapter- 11/lee_algo.py
from collections import deque

ROW = 5
COL = 9

class points:
    def __init__(self, x: int, y: int):
        self.x = x
        self.y = y

class Queue_Node:
    def __init__(self, pt: points, dist: int):
        self.pt = pt  
        self.dist = dist  


def Cell_isValid(row: int, col: int):
    return (row >= 0) and (row < ROW) and (col >= 0) and (col < COL)

rowNum = [-1, 0, 0, 1]
colNum = [0, -1, 1, 0]

def Lee_BFS(matrix, src: points, dest: points):
    if matrix[src.x][src.y] != 1 or matrix[dest.x][dest.y] != 1:
        return -1

    visited = [[False for i in range(COL)]
               for j in range(ROW)]
    visited[src.x][src.y] = True

    Q = deque()
    s = Queue_Node(src, 0)
    Q.append(s)

    while Q:
        curr = Q.popleft()
        pt = curr.pt
        if pt.x == dest.x and pt.y == dest.y:
            return curr.dist

        for i in range(4):
            row = pt.x + rowNum[i]
            col = pt.y + colNum[i]

            if (Cell_isValid(row, col) and
                    matrix[row][col] == 1 and
                    not visited[row][col]):
                visited[row][col] = True
                Adjcell = Queue_Node(points(row, col),
                                    curr.dist + 1)
                Q.append(Adjcell)

    return -1

def main():
    matrix = [[1, 0, 1, 1, 1, 1, 0, 1, 1, 1],
           [1, 0, 1, 0, 1, 1, 1, 0, 1, 1],
           [1, 1, 1, 0, 1, 1, 0, 1, 0, 1],
           [0, 0, 0, 0, 1, 0, 0, 0, 0, 1],
           [1, 1, 1, 0, 1, 1, 1, 0, 1, 0],
            [1, 0, 1, 1, 1, 1, 0, 1, 0, 0],
            [1, 0, 0, 0, 0, 0, 0, 0, 0, 1],
            [1, 0, 1, 1, 1, 1, 0, 1, 1, 1],
            [1, 1, 0, 0, 0, 0, 1, 0, 0, 1]]

    source = points(0, 0)
    dest = points(4,4)

    dist = Lee_BFS(matrix, source, dest)

    if dist != -1:
        print("Shortest Path is : ", dist)
    else:
        print("Shortest Path doesn't exist")

main()